tools/lkl/lib/fs.c

```diff
@@ -34,19 +37,81 @@ static uint32_t new_encode_dev(unsigned int major, unsigned int minor)
 	return (minor & 0xff) | (major << 8) | ((minor & ~0xff) << 12);
 }
 
+static int startswith(const char *str, const char *pre)
+{
+	return strncmp(pre, str, strlen(pre)) == 0;
+}
+
+static char *get_node_with_prefix(const char *path, const char *prefix,
+				  int *ret)
+{
+	struct lkl_dir *dir = NULL;
+	struct lkl_linux_dirent64 *dirent;
+	char *result = NULL;
+
+	dir = lkl_opendir(path, ret);
+	if (!dir)
+		return NULL;
+
+	while ((dirent = lkl_readdir(dir))) {
+		if (startswith(dirent->d_name, prefix))
+			result = strdup(dirent->d_name);
+	}
+	lkl_closedir(dir);
+
+	if (!result)
+		*ret = -LKL_ENOENT;
+
+	return result;
+}
+
 static long get_virtio_blkdev(int disk_id)
 {
-	char sysfs_path[] = "/sysfs/block/vda/dev";
+	char sysfs_path[LKL_PATH_MAX];
+	int sysfs_path_len = 0;
 	char buf[16] = { 0, };
 	long fd, ret;
 	int major, minor;
+	int opendir_ret;
+	char *virtio_name = NULL;
+	char *disk_name = NULL;
 
+	if (disk_id < 0)
+		return -LKL_EINVAL;
 
 	ret = lkl_mount_fs("sysfs");
 	if (ret < 0)
 		return ret;
 
-	sysfs_path[strlen("/sysfs/block/vd")] += disk_id;
+	if ((uint32_t) disk_id >= virtio_get_num_bootdevs())
+		ret = snprintf(sysfs_path, sizeof(sysfs_path), "/sysfs/devices/platform/virtio-mmio.%d.auto", disk_id - virtio_get_num_bootdevs());
```

Comment:
Since this line is very long (>100), you can split this one after the string, like:

ret = snprintf(sysfs_path, sizeof(sysfs_path), "/sysfs/devices/platform/virtio-mmio.%d.auto", 
                     disk_id - virtio_get_num_bootdevs());